File: solutions/elixir/food-chain/1/lib/food_chain.ex
defmodule FoodChain do

  @animals ["fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"]

  @exclamations [
    "It wriggled and jiggled and tickled inside her.",
    "How absurd to swallow a bird!",
    "Imagine that, to swallow a cat!",
    "What a hog, to swallow a dog!",
    "Just opened her throat and swallowed a goat!",
    "I don't know how she swallowed a cow!",
  ]
  @doc """
  Generate consecutive verses of the song 'I Know an Old Lady Who Swallowed a Fly'.
  """
  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, stop) do
    start..stop |> Enum.map(&verse/1) |> Enum.join("\n")
  end

  defp verse(number) do
    ["I know an old lady who swallowed a #{Enum.at(@animals, number-1)}.",
    comment(number),
    swallowing_phrase(number),
    last_sentence(number)
    ]
    |> List.flatten
    |> Enum.reject(&(String.length(&1) == 0))
    |> Enum.join("\n")
  end

  defp comment(number) when number < 2 or number >= 8, do: ""
  defp comment(number), do: Enum.at(@exclamations, number - 2)

  defp swallowing_phrase(number) when number < 2 or number >= 8, do: ""
  defp swallowing_phrase(number) do
      (number - 2)..0 |> Enum.map(fn n ->
        if n == 1 do
          "She swallowed the #{Enum.at(@animals, n+1)} to catch the spider that wriggled and jiggled and tickled inside her."
        else
          "She swallowed the #{Enum.at(@animals, n+1)} to catch the #{Enum.at(@animals, n)}."
        end
      end)
  end

  defp last_sentence(number) do
    if number == 8 do
      "She's dead, of course!\n"
    else
      "I don't know why she swallowed the fly. Perhaps she'll die.\n"
    end
  end

end

File: solutions/go/food-chain/1/food_chain.go
package foodchain

import (
    "fmt"
    "strings"
)

var animals = []string{
    "fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"}

var secondSentence = []string{
    "",
    "It wriggled and jiggled and tickled inside her.",
    "How absurd to swallow a bird!",
    "Imagine that, to swallow a cat!",
    "What a hog, to swallow a dog!",
    "Just opened her throat and swallowed a goat!",
    "I don't know how she swallowed a cow!"}

const lastSentence = "I don't know why she swallowed the fly. Perhaps she'll die."

const altLastSentence = "She's dead, of course!"

func Verse(v int) string {
	var out []string
    out = append(out,
                 fmt.Sprintf("I know an old lady who swallowed a %s.", animals[v-1]))
    if v < 8 {
        if v > 1 {
            out = append(out, secondSentence[v-1])
        }
        for i := 1; i < v; i++ {
            if animals[v-i-1] == "spider" {
                out = append(out,
                             fmt.Sprintf("She swallowed the %s to catch the %s that wriggled and jiggled and tickled inside her.",
                                         animals[v-i], animals[v-i-1]))
            } else {
            	out = append(out,
                             fmt.Sprintf("She swallowed the %s to catch the %s.",
                                         animals[v-i], animals[v-i-1]))
            }
        }
        out = append(out, lastSentence)
    } else {
        out = append(out, altLastSentence)
    }
    return strings.Join(out, "\n")
}

func Verses(start, end int) string {
	var out []string
    for i:= start; i <= end; i++ {
        out = append(out, Verse(i))
    }
    return strings.Join(out, "\n\n")
}

func Song() string {
	return Verses(1, 8)
}
